target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c

```diff
@@ -0,0 +1,484 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/module.h>
+#include <linux/of_platform.h>
+#include "i2c-rtl9300.h"
+
+#define REG(x)		(i2c->base + x + (i2c->scl_num ? 0x1c : 0))
+#define REG_MASK(clear, set, reg)	\
+			writel((readl(REG(reg)) & ~(clear)) | (set), REG(reg))
+
+struct i2c_drv_data {
+	int scl0_pin;
+	int scl1_pin;
+	int sda0_pin;
+	struct i2c_algorithm *algo;
+	int (*read)(struct rtl9300_i2c *i2c, u8 *buf, int len);
+	int (*write)(struct rtl9300_i2c *i2c, u8 *buf, int len);
+	void (*reg_addr_set)(struct rtl9300_i2c *i2c, u32 reg, u16 len);
+	int (*config_xfer)(struct rtl9300_i2c *i2c, u16 addr, u16 len);
+	int (*execute_xfer)(struct rtl9300_i2c *i2c, char read_write, int size,
+			    union i2c_smbus_data * data, int len);
+	void (*writel)(struct rtl9300_i2c *i2c, u32 data);
+	void (*config_io)(struct rtl9300_i2c *i2c, int scl_num, int sda_num);
+};
+			
+DEFINE_MUTEX(i2c_lock);
+
+static void rtl9300_i2c_reg_addr_set(struct rtl9300_i2c *i2c, u32 reg, u16 len)
+{
+	// Set register address width
+	REG_MASK(0x3 << RTL9300_I2C_CTRL2_MADDR_WIDTH, len << RTL9300_I2C_CTRL2_MADDR_WIDTH,
+		 RTL9300_I2C_CTRL2);
+
+	// Set register address
+	REG_MASK(0xffffff << RTL9300_I2C_CTRL1_MEM_ADDR, reg << RTL9300_I2C_CTRL1_MEM_ADDR,
+		 RTL9300_I2C_CTRL1);
+}
+
+static void rtl9310_i2c_reg_addr_set(struct rtl9300_i2c *i2c, u32 reg, u16 len)
+{
+	// Set register address width
+	REG_MASK(0x3 << RTL9310_I2C_CTRL_MADDR_WIDTH, len << RTL9310_I2C_CTRL_MADDR_WIDTH,
+		 RTL9310_I2C_CTRL);
+
+	// Set register address
+	writel(reg, REG(RTL9310_I2C_MEMADDR));
+}
+
+static void rtl9300_i2c_config_io(struct rtl9300_i2c *i2c, int scl_num, int sda_num)
+{
+	u32 v;
+
+	// Set SCL pin
+	REG_MASK(0, BIT(RTL9300_I2C_CTRL1_GPIO8_SCL_SEL), RTL9300_I2C_CTRL1);
+
+	// Set SDA pin
+	REG_MASK(0x7 << RTL9300_I2C_CTRL1_SDA_OUT_SEL,
+		 i2c->sda_num << RTL9300_I2C_CTRL1_SDA_OUT_SEL, RTL9300_I2C_CTRL1);
+
+	// Set SDA pin to I2C functionality
```

Comment:
This smells like it should be in a pin control driver, rather than an I2C device driver.

Reply:
You cannot do that because it needs to be dynamically controlled by the muxer, see: rtl9300_i2c_mux_select(), which sets RTL9300_I2C_CTRL1_GPIO8_SCL_SEL and RTL9300_I2C_CTRL1_SDA_OUT_SEL depending on the multiplex. The gpio pin is only for the time of the transfer actually an I2C pin and if you configure for a different multiplex this changes again.
But of course the original reason was that writing a pin control driver would be a huge overhead just for 2 pins controlled by this I2C driver. Maybe once we understand the SPI driver, there seem to be a couple of interesting SPI devices in the RTL93xx devices.

Reply:
Selecting the bus master for that pin makes no sense when it is configured as a GPIO pin. So selecting between GPIO and I2C functionality is a separate problem, and should be handled elsewhere. It's true that Realtek's register space is completely bonkers, but there are already some other bit(field)s in use that switch between GPIO or alternate function (e.g. the system LED pin).

Reply:
I've already experimented with a pinctrl driver for the switchcore register space, but I don't think I had figured out how to properly manage concurrent access to the registers that have these pin control bits.